File: java/Samples/Biometrics/VB/SimpleFingersSampleVB/EnrollFromImage.vb
Imports System
Imports System.IO
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Biometrics.Gui
Imports Neurotec.Images

Partial Public Class EnrollFromImage
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()

		OpenFileDialog.Filter = NImages.GetOpenFileFilterString(True, True)
		SaveFileDialog.Filter = NImages.GetSaveFileFilterString()
	End Sub

#End Region

#Region "Private fields"

	Private _image As NImage
	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _subjectFinger As NFinger

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub OnExtractionCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnExtractionCompleted), r)
		Else
			Dim status As NBiometricStatus = NBiometricStatus.None
			Try
				status = _biometricClient.EndCreateTemplate(r)
				If status <> NBiometricStatus.Ok Then
					MessageBox.Show(String.Format("Extraction failed. Status: {0}", status), Text, MessageBoxButtons.OK)
				End If
			Catch ex As Exception
				Utils.ShowException(ex)
			Finally
				If status = NBiometricStatus.Ok Then
					lblQuality.Text = String.Format("Quality: {0}", _subjectFinger.Objects.First().Quality)
					saveImageButton.Enabled = True
					saveTemplateButton.Enabled = True
					chbShowBinarizedImage.Enabled = True
				Else
					lblQuality.Text = String.Empty
					fingerView2.Finger = Nothing
					_subject = Nothing
					_subjectFinger = Nothing
				End If
			End Try
		End If
	End Sub

	Private Sub SetShownImage()
		fingerView2.ShownImage = If(chbShowBinarizedImage.Checked, ShownImage.Result, ShownImage.Original)
	End Sub

	Private Sub ExtractFeatures()
		If _image Is Nothing Then
			Return
		End If
		saveImageButton.Enabled = False
		saveTemplateButton.Enabled = False
		chbShowBinarizedImage.Enabled = False

		' Create a finger subject and begin extracting
		_subjectFinger = New NFinger With {.Image = _image}
		_subject = New NSubject()
		_subject.Fingers.Add(_subjectFinger)
		fingerView2.Finger = _subjectFinger
		SetShownImage()

		_biometricClient.BeginCreateTemplate(_subject, AddressOf OnExtractionCompleted, Nothing)
	End Sub

#End Region

#Region "Private form events"

	Private Sub OpenButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles openButton.Click
		openFileDialog.FileName = Nothing
		If openFileDialog.ShowDialog() = DialogResult.OK Then
			Try
				If _image IsNot Nothing Then
					_image.Dispose()
					_image = Nothing
				End If
				lblQuality.Text = String.Empty
				fingerView1.Finger = Nothing
				fingerView2.Finger = Nothing

				extractFeaturesButton.Enabled = False
				saveImageButton.Enabled = False
				saveTemplateButton.Enabled = False

				_image = NImage.FromFile(openFileDialog.FileName)
				Dim finger = New NFinger With {.Image = _image}
				fingerView1.Finger = finger
				extractFeaturesButton.Enabled = True
				ExtractFeatures()
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub ExtractFeaturesButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles extractFeaturesButton.Click
		ExtractFeatures()
	End Sub

	Private Sub DefaultButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles defaultButton.Click
		thresholdNumericUpDown.Value = 39
	End Sub

	Private Sub SaveImageButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles saveImageButton.Click
		If _subjectFinger IsNot Nothing Then
			saveFileDialog.FileName = String.Empty
			saveFileDialog.Filter = NImages.GetSaveFileFilterString()
			If saveFileDialog.ShowDialog() = DialogResult.OK Then
				Try
					If chbShowBinarizedImage.Checked Then
						_subjectFinger.BinarizedImage.Save(saveFileDialog.FileName)
					Else
						_subjectFinger.Image.Save(saveFileDialog.FileName)
					End If
				Catch ex As Exception
					Utils.ShowException(ex)
				End Try
			End If
		End If
	End Sub

	Private Sub ThresholdNumericUpDownValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles thresholdNumericUpDown.ValueChanged
		If _biometricClient IsNot Nothing Then
			_biometricClient.FingersQualityThreshold = CByte(thresholdNumericUpDown.Value)
			defaultButton.Enabled = True
		End If
	End Sub

	Private Sub SaveTemplateButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles saveTemplateButton.Click
		If _subject IsNot Nothing Then
			saveFileDialog.FileName = String.Empty
			saveFileDialog.Filter = String.Empty
			If saveFileDialog.ShowDialog() = DialogResult.OK Then
				Try
					File.WriteAllBytes(saveFileDialog.FileName, _subject.GetTemplateBuffer().ToArray())
				Catch ex As Exception
					Utils.ShowException(ex)
				End Try
			End If
		End If
	End Sub

	Private Sub EnrollFromImageVisibleChanged(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.VisibleChanged
		If Visible AndAlso _biometricClient IsNot Nothing Then
			thresholdNumericUpDown.Value = 39
			thresholdNumericUpDown.Enabled = True
			defaultButton.Enabled = True
			_biometricClient.FingersReturnBinarizedImage = True
		End If
	End Sub

	Private Sub ChbShowBinarizedImageCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chbShowBinarizedImage.CheckedChanged
		SetShownImage()
	End Sub

	Private Sub FingerView2MouseClick(ByVal sender As Object, ByVal e As MouseEventArgs) Handles fingerView2.MouseClick
		If e.Button = MouseButtons.Right AndAlso chbShowBinarizedImage.Enabled Then
			chbShowBinarizedImage.Checked = Not chbShowBinarizedImage.Checked
		End If
	End Sub

#End Region

End Class

File: java/Samples/Biometrics/CS/SimpleFingersSampleCS/EnrollFromImage.cs
using System;
using System.IO;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Biometrics.Gui;
using Neurotec.Images;

namespace Neurotec.Samples
{
	public partial class EnrollFromImage : UserControl
	{
		#region Public constructor

		public EnrollFromImage()
		{
			InitializeComponent();

			openFileDialog.Filter = NImages.GetOpenFileFilterString(true, true);
			saveFileDialog.Filter = NImages.GetSaveFileFilterString();
		}

		#endregion

		#region Private fields

		private NImage _image;
		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private NFinger _subjectFinger;

		private byte _defaultThreshold;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set 
			{
				_biometricClient = value;
				_defaultThreshold = _biometricClient.FingersQualityThreshold;
				thresholdNumericUpDown.Value = _defaultThreshold;
			}
		}

		#endregion

		#region Private methods

		private void OnExtractionCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnExtractionCompleted), r);
			}
			else
			{
				NBiometricStatus status = NBiometricStatus.None;
				try
				{
					status = _biometricClient.EndCreateTemplate(r);
					if (status != NBiometricStatus.Ok)
					{
						MessageBox.Show(string.Format(@"Extraction failed. Status: {0}", status), Text, MessageBoxButtons.OK);
					}
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
				finally
				{
					if (status == NBiometricStatus.Ok)
					{
						lblQuality.Text = string.Format("Quality: {0}", _subjectFinger.Objects.First().Quality);
						saveImageButton.Enabled = true;
						saveTemplateButton.Enabled = true;
						chbShowBinarizedImage.Enabled = true;
					}
					else
					{
						lblQuality.Text = string.Empty;
						fingerView2.Finger = null;
						_subject = null;
						_subjectFinger = null;
					}
				}
			}
		}

		private void ExtractFeatures()
		{
			if (_image == null) return;
			saveImageButton.Enabled = false;
			saveTemplateButton.Enabled = false;
			chbShowBinarizedImage.Enabled = false;

			// Create a finger subject and begin extracting
			_subjectFinger = new NFinger { Image = _image };
			_subject = new NSubject();
			_subject.Fingers.Add(_subjectFinger);
			fingerView2.Finger = _subjectFinger;

			_biometricClient.BeginCreateTemplate(_subject, OnExtractionCompleted, null);
		}

		#endregion

		#region Private form events

		private void OpenButtonClick(object sender, EventArgs e)
		{
			openFileDialog.FileName = null;
			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					if (_image != null)
					{
						_image.Dispose();
						_image = null;
					}
					lblQuality.Text = string.Empty;
					fingerView1.Finger = null;
					fingerView2.Finger = null;

					extractFeaturesButton.Enabled = false;
					saveImageButton.Enabled = false;
					saveTemplateButton.Enabled = false;

					_image = NImage.FromFile(openFileDialog.FileName);
					var finger = new NFinger { Image = _image };
					fingerView1.Finger = finger;
					extractFeaturesButton.Enabled = true;
					ExtractFeatures();
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void ExtractFeaturesButtonClick(object sender, EventArgs e)
		{
			ExtractFeatures();
		}

		private void DefaultButtonClick(object sender, EventArgs e)
		{
			thresholdNumericUpDown.Value = _defaultThreshold;
		}

		private void SaveImageButtonClick(object sender, EventArgs e)
		{
			if (_subjectFinger != null)
			{
				saveFileDialog.FileName = string.Empty;
				saveFileDialog.Filter = NImages.GetSaveFileFilterString();
				if (saveFileDialog.ShowDialog() == DialogResult.OK)
				{
					try
					{
						if (chbShowBinarizedImage.Checked)
						{
							_subjectFinger.BinarizedImage.Save(saveFileDialog.FileName);
						}
						else
						{
							_subjectFinger.Image.Save(saveFileDialog.FileName);
						}
					}
					catch (Exception ex)
					{
						Utils.ShowException(ex);
					}
				}
			}
		}

		private void ThresholdNumericUpDownValueChanged(object sender, EventArgs e)
		{
			if (_biometricClient != null)
			{
				_biometricClient.FingersQualityThreshold = (byte)thresholdNumericUpDown.Value;
				defaultButton.Enabled = true;
			}
		}

		private void SaveTemplateButtonClick(object sender, EventArgs e)
		{
			if (_subject != null)
			{
				saveFileDialog.FileName = string.Empty;
				saveFileDialog.Filter = string.Empty;
				if (saveFileDialog.ShowDialog() == DialogResult.OK)
				{
					try
					{
						File.WriteAllBytes(saveFileDialog.FileName, _subject.GetTemplateBuffer().ToArray());
					}
					catch (Exception ex)
					{
						Utils.ShowException(ex);
					}
				}
			}
		}

		private void EnrollFromImageVisibleChanged(object sender, EventArgs e)
		{
			if (Visible && _biometricClient != null)
			{
				thresholdNumericUpDown.Value = _defaultThreshold;
				thresholdNumericUpDown.Enabled = true;
				defaultButton.Enabled = true;
				_biometricClient.FingersReturnBinarizedImage = true;
			}
		}

		private void ChbShowBinarizedImageCheckedChanged(object sender, EventArgs e)
		{
			fingerView2.ShownImage = chbShowBinarizedImage.Checked ? ShownImage.Result : ShownImage.Original;
		}

		private void FingerView2MouseClick(object sender, MouseEventArgs e)
		{
			if (e.Button == MouseButtons.Right && chbShowBinarizedImage.Enabled)
			{
				chbShowBinarizedImage.Checked = !chbShowBinarizedImage.Checked;
			}
		}

		#endregion

	}
}
